### ISSUE TITLE
maybe add per-"block" phase to kernel `run()` cycle

### ISSUE THREAD

**Issue by warner** [Agoric/Agoric#4749]
## What is the Problem Being Solved?

@dtribble has suggested (for a while) that the kernel should know about block boundaries.

I've been resistant because I've felt that swingset shouldn't be aware about that aspect of the host app in which it's embedded. The kernel gets HostStorage (an object with the "SwingStore" API) from the host app, the host app tells the kernel when to run and for how long (via `runPolicy`), the kernel finishes, then the host app commits the state changes, and releases any outbound messages. In a consensus world (blockchains), the host includes the state changes in a block. In a solo world (`ag-solo`), the host must still commit changes before releasing messages (to prevent hangover inconsistency), but it's under no obligation to batch cranks in any particular way. A solo app would generally receive some input, process the kernel, commit+release, then go back to waiting for more input. It would only handle more than one input event at a time if there were many that arrived in a batch (one IO `onMessage` with multiple events inside it), or if it deliberately stalled in the hopes of batching events together (input Nagling).

The kernel is not the one telling the host when to end a block. The host provides a `runPolicy` for developer convenience (so that it can use `controller.run(policy)` instead of needing to have its own loop around `controller.step()`). So it's always seemed cleanest to me to keep the kernel out of the business of knowing about blocks, and leaving that entirely up to the host.

Howver, we now have some "maintenance" tasks which want to be run "soon", but which would be a waste to run multiple times in short succession. `dispatch.bringOutYourDead` is the primary example. If we know that we're going to do five deliveries to the same vat, it's wasteful to do BOYD after each one: we'd rather do it just once at the end. Currently, the kernel doesn't know when "the end" might come, so we use a counter on each vat: for every N deliveries, we do one BOYD. But we have no clear rationale on how to set this counter; we'll just have to run performance simulations to find a useful value.

Each BOYD delivery takes some fixed amount of time to run (probably `O(N)` in the number of live objects on the heap), and might manage to free up some RAM, and drop some imports (allowing other vats to do the same, later). BOYD is basically idempotent: running two in a row is unlikely to free anything the second time. Each `send` or `notify` delivery has the potential to allocate more memory, which then won't be freed until the next BOYD, so we'd like to run the BOYD soon, but not too soon.

I'm not sure this is enough to justify the work yet, but I'll use this ticket to collect use cases and potential designs.

## Description of the Design

If the kernel relies upon the host application to tell it about good times to run cleanup jobs, and the host doesn't do that, then the garbage could accumulate without bound. So a high-order decision is probably what exactly the host's responsibilities are. And I'd like an API that isn't inconvenient to use from a non-consensus application which doesn't have any notion of "block".

One possibility is that we define a "block" to be the set of cranks that fit into a `controller.run(policy)`. In this approach, the consensus app doesn't have to do anything special (it calls `controller.run(policy)` as usual). However, the kernel now has permission to do a non-trivial amount of cleanup work after the policy has said to stop. We'd have to make some judgement calls about how much work is allowed. Worst case, the kernel does a BOYD on every vat that received a message during the block. Each BOYD might cause the vat to emit drop/retire syscalls, which may cause drop/retire deliveries into other vats. To clean things up completely (i.e. spend time to buy less memory usage), we'd keep repeating this process until the last BOYD to each vat failed to emit any syscalls. Extensive cross-vat paths could allow this to go on a really long time, so we'd probably want to cut this off after two or three passes, however a particularly pathological reference graph (imagine a knitted sweater or shoelaces) might never be fully released if we impose a limit like this.

Given that our runPolicy is fairly advisory anyways (the policy is only consulted *after* each crank, so by the time the policy gets to say "stop", we've already exceeded some limit), this might not be so bad.

Using post-runPolicy cleanup also makes the solo case easier: solo nodes will probably just use `controller.run()` (i.e. drain the entire queue) after each input or timer event, so they wouldn't have to do anything special to allow the cleanup work to happen automatically.

Another possibility is that the host application is obligated to make a special `controller.finishWork()` call, after `controller.run()` or a long series of `controller.step()` calls. On one hand, this would allow the host application more flexibility: if it knew it was doing a really large amount of work, it could perform intermediate GC by calling `finishWork()` extra times (think `c.run(policy); c.finishWork(); c.run(morePolicy); c.finishWork(); commit();`). And if it knew it did a small amount of work but was going to do more very soon, it could deliberately *not* call `finishWork()` now, and only call it once every couple of blocks (Nagling the `finishWork()` calls). On the other hand, if the host app forgets, then garbage accumulates.

## Security Considerations

None that I can think of, but obviously determinimsm must be maintained: whatever the host app tells swingset, all members of a consensus kernel must do the same thing.

## Test Plan

unit tests

**Comment by warner** [comment]
@erights correctly pointed out that there's no strong reason this cleanup *needs* to be done by the end of a cosmos block. There's definitely a time-vs-space tradeoff involved, and the space consumption includes objects on other vats that we might be able to free. But if we deferred `bringOutYourDead` for an hour, it would just delay that space-reclamation by an hour.

The main argument for doing BOYD on each vat by the end of each block is a sort of "idle chains are fully clean" one. If we assume that the cost of DB usage is sort of integrated over time, then to minimize that cost metric, we'd want to delete garbage as quickly as possible. However, the marginal cost of doing the GC an hour later is probably small compared to the overall long-term storage required (you might say the waves are smaller than the depth of the ocean they ride upon), so in fact doing prompt GC might not save you all that much, as long as GC happens eventually.

I was also thinking that flushing the virtual-object LRU cache is an argument for doing this promptly (and for leaving vats in a clean state most of the time). Certainly we need this cache to be flushed before we can upgrade a vat and abandon the RAM-based state in its worker image. I'm intending to do a BOYD just before doing an upgrade, to ensure this. If we knew for certain that the vat was "clean", then we could avoid this call, but upgrade happens so infrequently that it's clearly better to do BOYD just before ugprade than to do it too frequently before those events.

I had been thinking that some sort of emergency upgrade would work better if vats were left in a clean state. I was imagining some kind of error condition (bad syscall, more likely a metering fault) that pauses/halts a vat, and we want to resume by upgrading the vat (to fix the problem), and that this would be easier if we knew the vat was clean. But now that I think about it, if the vat is in a bad state, we wouldn't be able to deliver a BOYD to it either, so in fact we don't have a way to guarantee the vat is clean that's any better than delivering the BOYD at the start of the upgrade.
